registry/consul/registry.go

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package consul
+
+import (
+	"strconv"
+)
+
+import (
+	consul "github.com/hashicorp/consul/api"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/registry"
+)
+
+func init() {
+	extension.SetRegistry("consul", newConsulRegistry)
+}
+
+// Consul registry wraps the consul client, to
+// register and subscribe service.
+type consulRegistry struct {
+	// Registry url.
+	*common.URL
+
+	// Consul client.
+	client *consul.Client
+
+	// Done field represents whether
+	// consul registry is closed.
+	done chan struct{}
+}
+
+func newConsulRegistry(url *common.URL) (registry.Registry, error) {
+	config := &consul.Config{Address: url.Location}
+	client, err := consul.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &consulRegistry{
+		URL:    url,
+		client: client,
+		done:   make(chan struct{}),
+	}
+
+	return r, nil
+}
+
+func (r *consulRegistry) Register(url common.URL) error {
+	var err error
+
+	role, _ := strconv.Atoi(r.URL.GetParam(constant.ROLE_KEY, ""))
+	if role == common.PROVIDER {
+		err = r.register(url)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
Why only register provider side? Where is consumer's registry?

Reply:
Because java version doesn't have consumer side register.

https://github.com/apache/dubbo/blob/98c46777581be7c8e6df18c6a3e55d3c470e19b4/dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java#L98-L105